File: src/main/java/com/tjh/riskfactor/repo/AnswerRepository.java
package com.tjh.riskfactor.repo;

import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;
import org.springframework.data.jpa.repository.JpaRepository;

import com.tjh.riskfactor.entity.form.Answer;
import com.tjh.riskfactor.entity.view.AnswerBrief;

import java.util.List;
import java.util.Collection;

@Repository
public interface AnswerRepository extends JpaRepository<Answer, Integer> {

    String ANSWER_BRIEF = "select a.id as id, c.username as creator, a.mtime as mtime from Answer a inner join a.task t on t.id = :taskId ";

    @Query(ANSWER_BRIEF + "inner join a.creator c on c.id = :uid")
    List<AnswerBrief> findTaskAnswersCreatedBy(Integer taskId, Integer uid);

    @Query(ANSWER_BRIEF + "inner join a.creator c")
    List<AnswerBrief> findAllByTaskId(Integer taskId);

    @Query("select a.id as id, c.username as creator, a.mtime as mtime from Answer a inner join a.task t on t.id = :taskId inner join a.creator c on c.username in :usernames")
    List<AnswerBrief> findAllByTaskIdCreatedBy(Integer taskId, Collection<String> usernames);

}

File: backend/src/main/kotlin/com/tjh/riskfactor/repo/AnswerRepository.kt
package com.tjh.riskfactor.repo

import org.springframework.stereotype.Repository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.jpa.repository.JpaRepository

import com.tjh.riskfactor.entity.form.Answer
import com.tjh.riskfactor.entity.view.AnswerView

private const val ANSWER_BRIEF = "select a.id as id, c.username as creator, a.mtime as mtime from Answer a inner join a.task t on t.id = :taskId "

@Repository
interface AnswerRepository: JpaRepository<Answer, Int> {

    @Query("select body from Answer where id = :id")
    fun bodyOf(id: Int): String?

    @Query(ANSWER_BRIEF + "inner join a.creator c on c.id = :uid")
    fun findTaskAnswersCreatedBy(taskId: Int, uid: Int): List<AnswerView>

    @Query(ANSWER_BRIEF + "inner join a.creator c")
    fun findAllByTaskId(taskId: Int): List<AnswerView>

    @Query(ANSWER_BRIEF + "inner join a.creator c on c.username in :usernames")
    fun findAllByTaskIdCreatedBy(taskId: Int, usernames: Collection<String>): List<AnswerView>

}
